File: Sources/SnickerStream/HzModClient.swift
import Foundation
import Network

/// HzMod streaming client (TCP). **Beta — not yet verified against real hardware.**
///
/// Protocol (from RattletraPM/Snickerstream `include/HzMod.au3`):
/// - Connect TCP to `<3DS>:6464`.
/// - Send, in order: a CPU-limit packet, a quality packet, then the stream-start packet.
/// - The 3DS then streams JPEG frames back over the same TCP connection (top screen only;
///   HzMod doesn't stream the bottom screen).
///
/// Frame framing in the original is a 13-byte-ish header (byte 0 = type `0x04`=JPEG, a
/// little-endian size, then reserved bytes) followed by the JPEG. Because those header
/// offsets are ambiguous in the source and we can't test against hardware, this client is
/// resilient: it scans the TCP byte stream for complete JPEGs (`FF D8` … `FF D9`) and emits
/// each one, ignoring the surrounding header bytes entirely.
final class HzModClient: StreamClient {
    private let config: StreamConfig
    private let queue = DispatchQueue(label: "snickerstream.hzmod.tcp")
    private var conn: NWConnection?
    private var buffer = Data()
    private var stopped = false

    private let soi = Data([0xFF, 0xD8])
    private let eoi = Data([0xFF, 0xD9])
    private let maxFrameBytes = 4 * 1024 * 1024

    var onFrame: ((Screen, Data) -> Void)?
    var onStatus: ((String) -> Void)?

    // Control packets (see HzMod.au3).
    private static let header: [UInt8] = [0x7E, 0x05, 0x00, 0x00]
    private func cpuLimitPacket(_ v: UInt8) -> Data { Data(Self.header + [0xFF, 0x00, 0x00, 0x00, v]) }
    private func qualityPacket(_ v: UInt8) -> Data { Data(Self.header + [0x03, 0x00, 0x00, 0x00, v]) }
    private static let startPacket = Data([0x7E, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01])

    init(config: StreamConfig) {
        self.config = config
    }

    func start() {
        stopped = false
        connect()
    }

    func stop() {
        queue.async { [self] in
            stopped = true
            onFrame = nil
            onStatus = nil
            conn?.cancel()
            conn = nil
            buffer.removeAll(keepingCapacity: false)
        }
    }

    func retry() {
        // HzMod has a single TCP connection; reconnect from scratch.
        queue.async { [self] in
            conn?.cancel()
            conn = nil
            buffer.removeAll(keepingCapacity: true)
            connect()
        }
    }

    func setQuality(_ quality: UInt8) {
        let q = min(100, max(1, quality))
        queue.async { [self] in
            conn?.send(content: qualityPacket(q), completion: .idempotent)
        }
    }

    func swapPriority() {
        // HzMod streams the top screen only — nothing to swap.
        onStatus?("HzMod streams the top screen only")
    }

    // MARK: - Connection

    private func connect() {
        guard let port = NWEndpoint.Port(rawValue: 6464) else { return }
        let connection = NWConnection(host: NWEndpoint.Host(config.ip), port: port, using: .tcp)
        self.conn = connection
        connection.stateUpdateHandler = { [weak self] state in
            guard let self = self else { return }
            switch state {
            case .ready:
                self.onStatus?("HzMod connected — starting stream")
                self.sendHandshake(on: connection)
                self.receive(on: connection)
            case .failed(let err):
                self.onStatus?("HzMod connection failed: \(err.localizedDescription)")
            case .cancelled:
                break
            default:
                break
            }
        }
        connection.start(queue: queue)
    }

    private func sendHandshake(on connection: NWConnection) {
        connection.send(content: cpuLimitPacket(config.cpuLimit), completion: .idempotent)
        connection.send(content: qualityPacket(min(100, max(1, config.quality))), completion: .idempotent)
        connection.send(content: Self.startPacket, completion: .idempotent)
    }

    private func receive(on connection: NWConnection) {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 64 * 1024) { [weak self] data, _, isComplete, error in
            guard let self = self, !self.stopped else { return }
            if let data = data, !data.isEmpty {
                self.buffer.append(data)
                self.extractFrames()
            }
            if error == nil && !isComplete {
                self.receive(on: connection)
            } else if isComplete {
                self.onStatus?("HzMod stream ended")
            }
        }
    }

    // MARK: - JPEG framing

    private func extractFrames() {
        while true {
            guard let soiRange = buffer.range(of: soi) else {
                // No frame start yet; keep at most a trailing byte (a split 0xFF).
                if buffer.count > 1 { buffer.removeSubrange(buffer.startIndex..<(buffer.endIndex - 1)) }
                return
            }
            // Discard any junk/header bytes before the JPEG start.
            if soiRange.lowerBound > buffer.startIndex {
                buffer.removeSubrange(buffer.startIndex..<soiRange.lowerBound)
            }
            // Look for the end-of-image after the start marker.
            let searchStart = buffer.index(buffer.startIndex, offsetBy: 2)
            guard let eoiRange = buffer.range(of: eoi, in: searchStart..<buffer.endIndex) else {
                // Frame still incomplete; guard against runaway growth on desync.
                if buffer.count > maxFrameBytes { buffer.removeAll(keepingCapacity: true) }
                return
            }
            let jpeg = buffer.subdata(in: buffer.startIndex..<eoiRange.upperBound)
            buffer.removeSubrange(buffer.startIndex..<eoiRange.upperBound)
            onFrame?(.top, jpeg)
        }
    }
}

File: Snickerstream4Win/Net/HzModClient.cs
using System.Net;
using System.Net.Sockets;

namespace Snickerstream4Win.Net;

/// <summary>
/// HzMod (beta) client. Connects over TCP/6464, sends CPU-limit, quality and start
/// packets, then receives JPEG frames over the same connection. The frame header is
/// unreliable, so we scan the byte stream for complete JPEGs (FF D8 … FF D9).
/// HzMod streams the top screen only.
/// </summary>
public sealed class HzModClient : IStreamClient
{
    public event Action<StreamFrame>? FrameReady;
    public event Action<string>? Status;
    public event Action<string>? Failed;
    public event Action? FirstFrame;

    private readonly string _ip;
    private int _quality;
    private readonly int _cpuLimit;

    private CancellationTokenSource? _cts;
    private Task? _worker;
    private TcpClient? _tcp;
    private volatile bool _firstFrameSeen;
    private volatile bool _stopped;
    private readonly object _writeLock = new();

    public HzModClient(string ip, int quality, int cpuLimit)
    {
        _ip = ip;
        _quality = Math.Clamp(quality, 1, 100);
        _cpuLimit = Math.Clamp(cpuLimit, 0, 255);
    }

    public static byte[] CpuLimitPacket(int cpuLimit) =>
        new byte[] { 0x7E, 0x05, 0x00, 0x00, 0xFF, 0x00, 0x00, 0x00, (byte)Math.Clamp(cpuLimit, 0, 255) };

    public static byte[] QualityPacket(int quality) =>
        new byte[] { 0x7E, 0x05, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00, (byte)Math.Clamp(quality, 1, 100) };

    public static byte[] StartPacket() =>
        new byte[] { 0x7E, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01 };

    public void Start()
    {
        _cts = new CancellationTokenSource();
        _worker = Task.Run(() => RunWithRetries(_cts.Token), _cts.Token);
    }

    private async Task RunWithRetries(CancellationToken token)
    {
        for (int attempt = 1; attempt <= 3 && !token.IsCancellationRequested && !_firstFrameSeen; attempt++)
        {
            Status?.Invoke($"Connecting… ({attempt}/3)");
            try
            {
                await Session(token);
            }
            catch (OperationCanceledException) { return; }
            catch (Exception ex)
            {
                Status?.Invoke($"HzMod error: {ex.Message}");
                await Task.Delay(800, token);
            }
            if (_firstFrameSeen) return;
        }

        if (!_firstFrameSeen && !token.IsCancellationRequested)
            Failed?.Invoke("No frames received via HzMod. Check the homebrew and the IP address.");
    }

    private async Task Session(CancellationToken token)
    {
        using var tcp = new TcpClient();
        _tcp = tcp;
        var connectTask = tcp.ConnectAsync(IPAddress.Parse(_ip), 6464);
        if (await Task.WhenAny(connectTask, Task.Delay(2500, token)) != connectTask)
            throw new TimeoutException("Timed out connecting to HzMod (6464).");
        await connectTask;

        var stream = tcp.GetStream();
        WritePacket(stream, CpuLimitPacket(_cpuLimit));
        WritePacket(stream, QualityPacket(_quality));
        WritePacket(stream, StartPacket());

        var scanner = new JpegStreamScanner();
        var rx = new byte[64 * 1024];

        while (!token.IsCancellationRequested)
        {
            int n;
            try { n = await stream.ReadAsync(rx, token); }
            catch (OperationCanceledException) { break; }
            if (n <= 0) break; // connection closed -> let retry loop reconnect

            scanner.Append(rx, n, jpeg =>
            {
                if (!_firstFrameSeen)
                {
                    _firstFrameSeen = true;
                    FirstFrame?.Invoke();
                }
                FrameReady?.Invoke(new StreamFrame(Screen.Top, jpeg));
            });
        }
    }

    private void WritePacket(NetworkStream stream, byte[] data)
    {
        lock (_writeLock)
        {
            stream.Write(data, 0, data.Length);
            stream.Flush();
        }
    }

    /// <summary>Live quality change: re-send the quality packet on the open connection.</summary>
    public void SetQuality(int quality)
    {
        _quality = Math.Clamp(quality, 1, 100);
        try
        {
            if (_tcp is { Connected: true })
                WritePacket(_tcp.GetStream(), QualityPacket(_quality));
        }
        catch { }
    }

    public void Stop()
    {
        if (_stopped) return;
        _stopped = true;
        try { _cts?.Cancel(); } catch { }
        try { _tcp?.Close(); } catch { }
    }

    public void Dispose() => Stop();
}

/// <summary>
/// Accumulates a TCP byte stream and emits each complete JPEG (start FF D8, end FF D9).
/// Tolerant of garbage between frames; bounded so a never-terminating frame can't grow
/// without limit.
/// </summary>
public sealed class JpegStreamScanner
{
    private readonly List<byte> _buf = new(128 * 1024);
    private const int MaxBuffer = 8 * 1024 * 1024;

    public void Append(byte[] data, int count, Action<byte[]> onJpeg)
    {
        for (int i = 0; i < count; i++) _buf.Add(data[i]);

        while (true)
        {
            int start = IndexOfMarker(0xD8, 0);
            if (start < 0)
            {
                // no SOI yet; keep only a trailing byte in case FF straddles reads
                if (_buf.Count > 1) _buf.RemoveRange(0, _buf.Count - 1);
                return;
            }
            if (start > 0) _buf.RemoveRange(0, start); // drop bytes before SOI

            int end = IndexOfMarker(0xD9, 2);
            if (end < 0)
            {
                if (_buf.Count > MaxBuffer) _buf.Clear(); // runaway guard
                return;
            }

            int len = end + 2; // include FF D9
            var jpeg = new byte[len];
            _buf.CopyTo(0, jpeg, 0, len);
            _buf.RemoveRange(0, len);
            onJpeg(jpeg);
        }
    }

    /// <summary>Finds an FF &lt;marker&gt; pair at or after startIndex.</summary>
    private int IndexOfMarker(byte marker, int startIndex)
    {
        for (int i = startIndex; i < _buf.Count - 1; i++)
            if (_buf[i] == 0xFF && _buf[i + 1] == marker)
                return i;
        return -1;
    }
}
